### PR TITLE
feat: make selector support expression selectors.

### LINKED ISSUE #1266: selector support set-based requirements
## Feature Request

**Describe the feature you'd like:**

Now Chaos Mesh support define the scope of chaos experiments by [selector](https://chaos-mesh.org/docs/user_guides/experiment_scope). But the selector don't support set-based, which is similar with [resources-that-support-set-based-requirements](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#resources-that-support-set-based-requirements).

### PR DESCRIPTION
**Pull request by Colstuwjx**
### What problem does this PR solve?

Fixed #1266 

### What is changed and how does it work?
Add expressionSelectors field in selector spec and use metav1.LabelSelectors to do expression match operations.

### Checklist

Tests

- [ ] Unit test
- [ ] E2E test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Breaking backward compatibility

Related changes

- [ ] Need to update the documentation

### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by Colstuwjx** [comment]
It's still WIP, I need also add some testcases for testing it. But I think it will be ready for review very soon.

**Comment by Colstuwjx** [comment]
CI would throw error if I keep using `metav1.LabelSelectorRequirement`:

```
Please make check before creating a PR
diff --git a/api/v1alpha1/zz_generated.deepcopy.go b/api/v1alpha1/zz_generated.deepcopy.go
index 6c3a89e..a95345d 100644
--- a/api/v1alpha1/zz_generated.deepcopy.go
+++ b/api/v1alpha1/zz_generated.deepcopy.go
@@ -18,7 +18,7 @@
 package v1alpha1
 
 import (
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 )
 
@@ -1731,7 +1731,7 @@ func (in *SelectorSpec) DeepCopyInto(out *SelectorSpec) {
 	}
 	if in.ExpressionSelectors != nil {
 		in, out := &in.ExpressionSelectors, &out.ExpressionSelectors
-		*out = make([]metav1.LabelSelectorRequirement, len(*in))
+		*out = make([]v1.LabelSelectorRequirement, len(*in))
 		for i := range *in {
 			(*in)[i].DeepCopyInto(&(*out)[i])
 		}
Error: Process completed with exit code 1.
```

So I choose workaround by:

```
// LabelSelectorRequirements is list of LabelSelectorRequirement
type LabelSelectorRequirements []metav1.LabelSelectorRequirement
```

Any better solution will be appreciated.

**Comment by STRRL** [comment]
Hi @cwen0 @WangXiangUSTC , how about making a broken change for the label selector?

The `LabelSelector` in Kubernetes is mostly like:

```go
// A label selector is a label query over a set of resources. The result of matchLabels and
// matchExpressions are ANDed. An empty label selector matches all objects. A null
// label selector matches no objects.
type LabelSelector struct {
	// matchLabels is a map of {key,value} pairs. A single {key,value} in the matchLabels
	// map is equivalent to an element of matchExpressions, whose key field is "key", the
	// operator is "In", and the values array contains only "value". The requirements are ANDed.
	// +optional
	MatchLabels map[string]string `json:"matchLabels,omitempty" protobuf:"bytes,1,rep,name=matchLabels"`
	// matchExpressions is a list of label selector requirements. The requirements are ANDed.
	// +optional
	MatchExpressions []LabelSelectorRequirement `json:"matchExpressions,omitempty" protobuf:"bytes,2,rep,name=matchExpressions"`
}
```

In Choas Mesh repo,  after we introduce the expressions selector, there are two fields in the CRD.
I think it's better to keep the same as kubernetes.

**Comment by Colstuwjx** [comment]
@STRRL Did you mean the following structure?

```
type SelectorSpec struct {
	// Namespaces is a set of namespace to which objects belong.
	// +optional
	Namespaces []string `json:"namespaces,omitempty"`

	// Nodes is a set of node name and objects must belong to these nodes.
	// +optional
	Nodes []string `json:"nodes,omitempty"`

	// Pods is a map of string keys and a set values that used to select pods.
	// The key defines the namespace which pods belong,
	// and the each values is a set of pod names.
	// +optional
	Pods map[string][]string `json:"pods,omitempty"`

	// Map of string keys and values that can be used to select nodes.
	// Selector which must match a node's labels,
	// and objects must belong to these selected nodes.
	// +optional
	NodeSelectors map[string]string `json:"nodeSelectors,omitempty"`

	// Map of string keys and values that can be used to select objects.
	// A selector based on fields.
	// +optional
	FieldSelectors map[string]string `json:"fieldSelectors,omitempty"`

	// A selector based on labels.
	// +optional
	LabelSelector metav1. LabelSelector `json:"labelSelectors,omitempty"`
       ...
}
```

**Comment by STRRL** [comment]
> Did you mean the following structure? Directly using metav1.LabelSelector?

Yes.

**Comment by WangXiangUSTC** [comment]
> Hi @cwen0 @WangXiangUSTC @YangKeao @Gallardot @Colstuwjx , how about making a broken change for the label selector?
> 
> The `LabelSelector` in Kubernetes is mostly like:
> 
> ```go
> // A label selector is a label query over a set of resources. The result of matchLabels and
> // matchExpressions are ANDed. An empty label selector matches all objects. A null
> // label selector matches no objects.
> type LabelSelector struct {
> 	// matchLabels is a map of {key,value} pairs. A single {key,value} in the matchLabels
> 	// map is equivalent to an element of matchExpressions, whose key field is "key", the
> 	// operator is "In", and the values array contains only "value". The requirements are ANDed.
> 	// +optional
> 	MatchLabels map[string]string `json:"matchLabels,omitempty" protobuf:"bytes,1,rep,name=matchLabels"`
> 	// matchExpressions is a list of label selector requirements. The requirements are ANDed.
> 	// +optional
> 	MatchExpressions []LabelSelectorRequirement `json:"matchExpressions,omitempty" protobuf:"bytes,2,rep,name=matchExpressions"`
> }
> ```
> 
> In Choas Mesh repo, after we introduce the expressions selector, there are two fields in the CRD.
> I think it's better to keep the same as kubernetes.

I think it's better to support these two fields recently, and deprecate the old one in the next big version(maybe 2.0?)

**Review by STRRL — APPROVED** [review]
LGTM

**Comment by WangXiangUSTC** [comment]
@Colstuwjx ci failed, please fix it

**Comment by Colstuwjx** [comment]
@STRRL @WangXiangUSTC PTAL!

File: pkg/core/experiment.go [hunk 1]
```diff
@@ -134,6 +137,10 @@ func (s *SelectorInfo) ParseSelector() v1alpha1.SelectorSpec {
 		selector.LabelSelectors[key] = val
 	}
 
+	for _, expr := range s.ExpressionSelectors {
+		selector.ExpressionSelectors = append(selector.ExpressionSelectors, expr)
+	}
```
Comment (WangXiangUSTC) [inline pkg/core/experiment.go:142]: how about using `selector.ExpressionSelectors = append(selector.ExpressionSelectors, s.ExpressionSelectors...)`

File: pkg/apivalidator/scope_validator.go [hunk 2]
```diff
@@ -17,7 +17,9 @@ import (
 	"strconv"
 
 	"github.com/go-playground/validator/v10"
+
```
Comment (WangXiangUSTC) [inline pkg/apivalidator/scope_validator.go:20]: can delete this line

**Review by WangXiangUSTC — COMMENTED** [review]
rest LGTM

**Review by WangXiangUSTC — APPROVED** [review]
LGTM

**Review by STRRL — APPROVED** [review]
LGTM

**Pull request merged**